### PR TITLE
UI Breadcrumbs

### PR DESCRIPTION
**Pull request by nhaagen**
This is a Breadcrumbs-implementation for UI making use of UILinks.

### CODE REVIEW

**Comment by alex40724** [comment]
Looks good to me.

File: src/UI/Factory.php [hunk 1]
```diff
@@ -538,4 +538,33 @@ public function link();
...
+	 * context: >
+	 *
+	 * rules:
+	 *   usage:
+	 *     1: Crumbs MUST trigger navigation.
+	 * ---
+	 * @param 	\ILIAS\UI\Component\Link\Standard[] 	$crumbs 	a list of Links
+	 * @return 	\ILIAS\UI\Component\Breadcrumbs\Breadcrumbs
+	 **/
+	public function breadcrumbs($crumbs);
```
Comment (klees) [inline src/UI/Factory.php]: Hint `array` here.

File: src/UI/Factory.php [hunk 2]
```diff
@@ -538,4 +538,33 @@ public function link();
...
+	 *     Breadcrumbs-entries are rendered as horizontally arranged UI Links
+	 *     with a seperator in-between.
+	 *   effect: >
+	 *     Clicking on an entry will get the user to the respective location.
+	 *
+	 * context: >
+	 *
+	 * rules:
+	 *   usage:
+	 *     1: Crumbs MUST trigger navigation.
```
Comment (klees) [inline src/UI/Factory.php]: navigation = page change?

**Comment by Amstutz** [comment]
LGTM in general. 

However, please move the content of https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-2/templates/default/less/Services/Locator/delos.less next to the template. Also refactor it slightly:
```
/* Services/Locator */

ol.breadcrumb {
	margin: 0;
	font-weight: 600; <-- introduce a new breadcrum variable for this.
	padding-left: 15px; <-- use @padding-large-horizontal
	font-size: .8em; <-- use @font-size-base or a variation of is.
}

[dir="rtl"] ol.breadcrumb {
	padding-right: 15px; <-- use @padding-large-horizontal
}

.breadcrumb a {
/*	color: #a0a0a0; */ <-- Maybe delete this block?
}

.breadcrumb > li + li:before {
/*	 color: #a0a0a0; */ <-- Maybe delete this block?
 }
```

**Review by Amstutz — CHANGES_REQUESTED** [review]
Please fix conflicts, run tests again and slightly refactor less.

**Review by Amstutz — APPROVED** [review]
Thank you very much.

**Comment by nhaagen** [comment]
You're most welcome. Will you hit the merge-button?

**Comment by Amstutz** [comment]
Normally you could as soon as this is accepted. However in this particular case, I just realised we also need the JF approval for this (thx @klees). This is already done for Icons und Tags... So we discuss this on monday and you can merge that on monday evening if accepted by JF.

**Comment by nhaagen** [comment]
Thanks; however, I cannot merge (also Icons/Tags), since I have no write access to ILIAS...

File: src/UI/Factory.php [hunk 3]
```diff
@@ -606,4 +606,33 @@ public function item();
 	 **/
 	public function icon();
 
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     Breadcrumbs is a type of secondary navigation scheme. It eases the
```
Comment (atoedt) [inline src/UI/Factory.php]: Breadcrumbs is a supplemental navigation scheme.

File: src/UI/Factory.php [hunk 4]
```diff
@@ -606,4 +606,33 @@ public function item();
...
+	 *     user's navigation back to higher items in hierarchical (or prior items
```
Comment (atoedt) [inline src/UI/Factory.php]: Users user's can backtrack the path they have taken in the tree. The backtracking is supported for hierarchical and temporal structures.

File: src/UI/Factory.php [hunk 5]
```diff
@@ -606,4 +606,33 @@ public function item();
...
+	 *     in chronological) structures.
+	 *     Breadcrumbs also serve as an effective visual aid indicating the
```
Comment (atoedt) [inline src/UI/Factory.php]: This sentence must go if the Beardcrums is conceptualized to encompass the "prior items". They are no way an effective visual aide to locate anything. They offer the opportuntiy to backtrack but they do not help locating anything.

File: src/UI/Factory.php [hunk 6]
```diff
@@ -606,4 +606,33 @@ public function item();
...
+	 *     user's location on a website.
+	 *     The term is based on the story of Hansel and Gretel and is well established,
```
Comment (atoedt) [inline src/UI/Factory.php]: The Adriadne and Grimms sentence does not add specificity. Please remove.

File: src/UI/Factory.php (hunk 4 above)
Comment (atoedt) [inline src/UI/Factory.php]: Please think long and hard whether to include "prior items" or not

File: src/UI/Factory.php [hunk 7]
```diff
@@ -606,4 +606,33 @@ public function item();
...
+	 *     so we will stick to it, although in the context of ILIAS "Ariadne's Path"
+	 *     would probably be more appropriate.
+	 *   composition: >
+	 *     Breadcrumbs-entries are rendered as horizontally arranged UI Links
```
Comment (atoedt) [inline src/UI/Factory.php]: does not apply to prior items 
consider dropping the horizontally or adding a vertically

File: src/UI/Factory.php [hunk 8]
```diff
@@ -606,4 +606,33 @@ public function item();
...
+	 *     with a seperator in-between.
+	 *   effect: >
+	 *     Clicking on an entry will get the user to the respective location.
+	 *
+	 * context: >
+	 *
+	 * rules:
+	 *   usage:
+	 *     1: Crumbs MUST trigger navigation.
+	 * ---
```
Comment (atoedt) [inline src/UI/Factory.php]: Breadcrums MUST NOT entail objects users have no Visible permission for.

File: src/UI/Component/Breadcrumbs/Breadcrumbs.php [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * @package ILIAS\UI\Component\Breadcrumbs
+ */
+interface Breadcrumbs extends \ILIAS\UI\Component\Component {
+
+	/**
+	 * Get all crumbs.
+	 *
+	 * @return 	\ILIAS\UI\Component\Link\Standard[]
+	 */
+	public function getCrumbs();
```
Comment (klees) [inline src/UI/Component/Breadcrumbs/Breadcrumbs.php]: `getEntries` please.

**Review by klees — COMMENTED** [review]
LGTM in general, minor wording issue.

File: src/UI/Factory.php (hunk 2 above)
Comment (atoedt) [reply src/UI/Factory.php]: what is a page? 
Change from object to object, from reporitory to PD or Administration
Minimally from one sub-item of an Objects to another or from one sub-tab to another

File: src/UI/Factory.php (hunk 4 above)
Comment (matthiaskunkel) [reply src/UI/Factory.php]: Jour Fixe: please remove "(or prior items in chronological)"

Comment (matthiaskunkel) [reply src/UI/Factory.php]: Please remove "back" from the description.

File: src/UI/Factory.php (hunk 5 above)
Comment (matthiaskunkel) [reply src/UI/Factory.php]: Rejected

File: src/UI/Factory.php (hunk 7 above)
Comment (matthiaskunkel) [reply src/UI/Factory.php]: rejected

File: src/UI/Factory.php [hunk 10]
```diff
@@ -606,4 +606,33 @@ public function item();
...
+	 *     The term is based on the story of Hansel and Gretel and is well established,
+	 *     so we will stick to it, although in the context of ILIAS "Ariadne's Path"
+	 *     would probably be more appropriate.
+	 *   composition: >
+	 *     Breadcrumbs-entries are rendered as horizontally arranged UI Links
+	 *     with a seperator in-between.
+	 *   effect: >
+	 *     Clicking on an entry will get the user to the respective location.
+	 *
+	 * context: >
```
Comment (matthiaskunkel) [inline src/UI/Factory.php:636]: Please add the following contextes:
- search results
- info page
- the no-name space under the main menu

File: src/UI/Factory.php (hunk 2 above)
Comment (matthiaskunkel) [reply src/UI/Factory.php]: Jour Fixe: "Crumbs must trigger navigation to other resources of the system." (similar to Link definition)

File: src/UI/Factory.php (hunk 8 above)
Comment (klees) [reply src/UI/Factory.php]: "Breadcrumbs MUST only entail resources that are visible to the user."

File: src/UI/Component/Breadcrumbs/Breadcrumbs.php (hunk 9 above)
Comment (klees) [reply src/UI/Component/Breadcrumbs/Breadcrumbs.php]: Please rename the entries to `items` completely, as we use this wording in the rules.

File: src/UI/Factory.php (hunk 8 above)
Comment (matthiaskunkel) [reply src/UI/Factory.php]: Jour Fixe: We skip this as it is only related to repository object.

File: src/UI/Component/Breadcrumbs/Breadcrumbs.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+	/**
+	 * Append an crumb-entry to the bar.
+	 *
+	 * @param 	\ILIAS\UI\Component\Link\Standard 	$crumb
+	 * @return 	Breadcrumbs
+	 */
+	public function withAppendedEntry($crumb);
```
Comment (matthiaskunkel) [inline src/UI/Component/Breadcrumbs/Breadcrumbs.php]: Jour Fixe: Rename to 'withAppendedItem'

**Comment by matthiaskunkel** [comment]
Jour Fixe, July 03, 2017: Thank you for this pull request. We have added a few wording improvements. PR is appreciated and can be merged to trunk.

**Comment by klees** [comment]
Hi @nhaagen, please add the context mentioned in the open comment from @matthiaskunkel, solve conflicts and then contact me immediately, so I can push the "Merge"-button.

**Pull request merged**